### PR TITLE
add blake3

### PR DESCRIPTION
**Pull request by haneefmubarak**
blake3 is a secure and insanely fast cryptographic function

### CODE REVIEW

**Comment by waruqi** [comment]
please push to dev branch next.

File: packages/b/blake3/xmake.lua [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+                    elseif is_plat("windows") then
+                        add_files("c/*x86-64_windows_msvc.S")
+                    else
+                        add_files("c/*x86-64_unix.S")
+                    end
+                elseif is_arch("arm.*") then
+                    add_files("c/blake3_neon.c")
+
+                    if is_arch("arm64") then
+                        add_cflags("-DBLAKE3_USE_NEON=1")
```
Comment (waruqi) [inline packages/b/blake3/xmake.lua:32]: add_defines

File: packages/b/blake3/xmake.lua [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+             "https://github.com/BLAKE3-team/BLAKE3.git")
+    add_versions("1.3.1", "112becf0983b5c83efff07f20b458f2dbcdbd768fd46502e7ddd831b83550109")
+
+    on_install(function (package)
+        local configs = {}
+        io.writefile("xmake.lua", [[
+            add_rules("mode.release", "mode.debug")
+            target("blake3")
+                set_kind("$(kind)")
+                add_cflags("-O3") -- as advised by BLAKE3-team in README (cryptography)
```
Comment (waruqi) [inline packages/b/blake3/xmake.lua:16]: release.mode will set it, please remove it

**Comment by haneefmubarak** [comment]
sounds good, fixed both and pushed

**Comment by haneefmubarak** [comment]
pushed a fix for failing tests

File: packages/b/blake3/xmake.lua [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    on_install(function (package)
+        local configs = {}
+        io.writefile("xmake.lua", [[
+            add_rules("mode.release", "mode.debug")
+            target("blake3")
+                set_kind("$(kind)")
+                add_files("c/blake3.c", "c/blake3_dispatch.c", "c/blake3_portable.c")
+                add_headerfiles("c/blake3.h")
+
+                if is_arch("x86_64") then
```
Comment (waruqi) [inline packages/b/blake3/xmake.lua:19]: if is_arch("x86_64", "x64") then

File: packages/b/blake3/xmake.lua [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+                    end
+                elseif is_arch("arm.*") then
+                    -- only used to build the NEON accelerated routines;
+                    -- correct routine is detected and dispatched at runtime
+                    add_cflags("-mfpu=neon-vfpv4", "-mfloat-abi=hard")
+                    add_files("c/blake3_neon.c")
+
+                    if is_arch("arm64") then
+                        add_defines("BLAKE3_USE_NEON=1")
+                    end
```
Comment (waruqi) [inline packages/b/blake3/xmake.lua:35]:
```
                elseif is_arch("arm64", "arm64-v8a") then
                    add_cflags("-mfpu=neon", "-mfloat-abi=soft")
                    add_files("c/blake3_neon.c")
                    add_defines("BLAKE3_USE_NEON=1")
```

**Comment by waruqi** [comment]
```lua

    on_install(function (package)
        local configs = {}
        io.writefile("xmake.lua", [[
            add_rules("mode.release", "mode.debug")
            target("blake3")
                set_kind("$(kind)")
                add_files("c/blake3.c", "c/blake3_dispatch.c", "c/blake3_portable.c")
                add_headerfiles("c/blake3.h")
                if is_arch("x86_64", "x64") then
                    if is_subhost("msys", "cygwin") then
                        add_files("c/*x86-64_windows_gnu.S")
                    elseif is_plat("windows") then
                        add_files("c/*x86-64_windows_msvc.asm")
                    else
                        add_files("c/*x86-64_unix.S")
                    end
                elseif is_arch("arm64", "arm64-v8a") then
                    add_cflags("-mfpu=neon", "-mfloat-abi=soft")
                    add_files("c/blake3_neon.c")
                    add_defines("BLAKE3_USE_NEON=1")
                end
        ]])
        if package:config("shared") then
            configs.kind = "shared"
        end
        import("package.tools.xmake").install(package, configs)
    end)
```

File: packages/b/blake3/xmake.lua (hunk 3 above)
Comment (haneefmubarak) [reply packages/b/blake3/xmake.lua:19]: Gotcha, am applying all of your changes. For my own learning though, what is the difference between x86_64 and x64 supposed to be?

**Comment by haneefmubarak** [comment]
@oconnor663 do you have any idea why this is failing to link on Windows x64? I tried to adapt the directions at [BLAKE3/c/README.md#building](https://github.com/BLAKE3-team/BLAKE3/blob/master/c/README.md#building), but I can't seem to get the linking to work correctly.

File: packages/b/blake3/xmake.lua (hunk 3 above)
Comment (waruqi) [reply packages/b/blake3/xmake.lua:19]: x86 x64 for windows/msvc

**Comment by waruqi** [comment]
> @oconnor663 do you have any idea why this is failing to link on Windows x64? I tried to adapt the directions at [BLAKE3/c/README.md#building](https://github.com/BLAKE3-team/BLAKE3/blob/master/c/README.md#building), but I can't seem to get the linking to work correctly.

x64

**Comment by haneefmubarak** [comment]
I think it seems to fail even with that (772fed7)

**Comment by oconnor663** [comment]
I'd want to drill into exactly what `add_files("c/*x86-64_windows_msvc.asm")` is doing. The linker errors I see like these...

```
blake3.lib(blake3_dispatch.c.obj) : error LNK2019: unresolved external symbol _blake3_compress_in_place_sse2 referenced in function _blake3_compress_in_place
blake3.lib(blake3_dispatch.c.obj) : error LNK2019: unresolved external symbol _blake3_compress_xof_sse2 referenced in function _blake3_compress_xof
blake3.lib(blake3_dispatch.c.obj) : error LNK2019: unresolved external symbol _blake3_hash_many_sse2 referenced in function _blake3_hash_many
blake3.lib(blake3_dispatch.c.obj) : error LNK2019: unresolved external symbol _blake3_compress_in_place_sse41 referenced in function _blake3_compress_in_place
blake3.lib(blake3_dispatch.c.obj) : error LNK2019: unresolved external symbol _blake3_compress_xof_sse41 referenced in function _blake3_compress_xof
blake3.lib(blake3_dispatch.c.obj) : error LNK2019: unresolved external symbol _blake3_hash_many_sse41 referenced in function _blake3_hash_many
blake3.lib(blake3_dispatch.c.obj) : error LNK2019: unresolved external symbol _blake3_hash_many_avx2 referenced in function _blake3_hash_many
blake3.lib(blake3_dispatch.c.obj) : error LNK2019: unresolved external symbol _blake3_compress_in_place_avx512 referenced in function _blake3_compress_in_place
blake3.lib(blake3_dispatch.c.obj) : error LNK2019: unresolved external symbol _blake3_compress_xof_avx512 referenced in function _blake3_compress_xof
blake3.lib(blake3_dispatch.c.obj) : error LNK2019: unresolved external symbol _blake3_hash_many_avx512 referenced in function _blake3_hash_many
```

...are all for symbols defined in those `.asm` files. So one way or another, it seems like that `add_files` call might be doing nothign. Is it possible to get a lot of the exact compiler/linker commands that end up running? I've seen issues building Python modules before where GCC on Linux was happy to accept .S files along with regular .c sources, but MSVC got confused and upset. I ended up having to [shell out to the assembler explicitly on Windows](https://github.com/oconnor663/blake3-py/blob/72fd6c63917bf4e296e39a82cc25e6e003534370/c_impl/setup.py#L195-L204). Very annoying :(

**Comment by waruqi** [comment]
> I think it seems to fail even with that ([772fed7](https://github.com/xmake-io/xmake-repo/commit/772fed7f3c2520f3aa2e783fc8e0ac1001512cce))

It will fix link errors for static library.

> I'd want to drill into exactly what add_files("c/*x86-64_windows_msvc.asm") is doing. The linker errors I see like these...

you need export symbols for windows shared library, otherwise `xxx.lib` import library will be not generated.

you can use `add_rules("utils.symbols.export_all")` to export all, or modify code to export them.

```lua
      elseif is_plat("windows") then
          add_files("c/*x86-64_windows_msvc.asm")
          if is_kind("shared") then
              add_rules("utils.symbols.export_all")
          end
```

it works for me.

image

**Comment by waruqi** [comment]
Also, since no asm file has been added on the x86 platform, it needs to be fixed as well. There are now two different asm files on the ci.

- [x] x64/static link errors (fixed, `is_arch("x64")`)
- [ ] x64 shared link errors, you need export symbols
- [ ] x86 static/shared link errors, missing asm files

**Pull request merged**